### DISCUSSION TITLE
Detect USB plugged in?

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12175] [ESP32]
Apparently the esp32 has a register that detects if the usb is plugged in? https://www.esp32.com/viewtopic.php?t=26380 and https://www.esp32.com/viewtopic.php?t=31293  Just wondering if it's possible to read it from upython? Maybe an atomic_bitmask _read() or something?

**Comment by karfas** [comment]
What are you trying to do ?
The ESP32 has no USB support at all. The ESP32-S3/C3/?? variants **have** some USB support; you might be able to find the details (registers, etc) in the datasheets.

**Comment by rkompass** [comment]
You seem to refer to this piece of code:
```C
int is_plugged_usb(void){
    uint32_t *aa = USB_SERIAL_JTAG_FRAM_NUM_REG;
    uint32_t first = *aa;
    vTaskDelay(pdMS_TO_TICKS(10));
    return (int) (*aa - first);
}
```
which refers to C code headers included that are part of the IDF  that are specific for each esp device.
Searching through the code for the ESP32-C3 gave:
`#define DR_REG_USB_SERIAL_JTAG_BASE             0x60043000`  and
`#define USB_SERIAL_JTAG_FRAM_NUM_REG (DR_REG_USB_SERIAL_JTAG_BASE + 0x24)`.
If you use a different device then better repeat the search yourself in the IDF.

So you have an address where to look at: `0x60043024` and you can read the value there with
```py
from machine import mem32
first = mem32[0x60043024]
```
then perhaps use `delay_ms(10)` from the time module and so on.

**Reply by kjm1102** [reply]
Am I likely to find this memory location for something like the S2 or does it only exist in USB aware variants like the C3?

**Reply by kjm1102** [reply]
Maybe I should stick with the gpio jumper. <img width="547" alt="2308esp32" src="https://github.com/micropython/micropython/assets/92007731/be222537-2071-49e1-a20a-02068a2bb7f8">

**Reply by rkompass** [reply]
> Am I likely to find this memory location for something like the S2

USB_SERIAL_JTAG_FRAM_NUM_REG occurs in include files for the C3, C6, H2, P4 and S3 variants.

**Reply by karfas** [reply]
See the [S2 technical reference](https://www.espressif.com/sites/default/files/documentation/esp32-s2_technical_reference_manual_en.pdf).
There they describe registers and addresses für the USB OTG interface.
